codec/reflectcodec/type_codec.go

```diff
@@ -332,6 +349,70 @@ func (c *genericCodec) marshal(value reflect.Value, p *wrappers.Packer, maxSlice
 				return err
 			}
 		}
+		return nil
+	case reflect.Map:
+		keys := value.MapKeys()
+		numElts := len(keys)
+		if uint32(numElts) > maxSliceLen {
+			return fmt.Errorf("%w; slice length, %d, exceeds maximum length, %d",
+				codec.ErrMaxSliceLenExceeded,
+				numElts,
+				maxSliceLen,
+			)
+		}
+		p.PackInt(uint32(numElts)) // pack # elements
+		if p.Err != nil {
+			return p.Err
+		}
+
+		type keyTuple struct {
+			key    reflect.Value
+			binary []byte
+		}
+
+		sortedKeys := make([]keyTuple, 0)
+
+		for _, key := range keys {
+			p := wrappers.Packer{
+				MaxSize: int(c.maxSliceLen),
+				Bytes:   make([]byte, 0, 128),
+			}
+			if err := c.marshal(key, &p, c.maxSliceLen); err != nil {
+				return err
+			}
+			sortedKeys = append(sortedKeys, keyTuple{
+				key:    key,
+				binary: p.Bytes,
+			})
+		}
+
+		sort.Slice(sortedKeys, func(a, b int) bool {
+			minLength := len(sortedKeys[a].binary)
+			if minLength > len(sortedKeys[b].binary) {
+				minLength = len(sortedKeys[b].binary)
+			}
+
+			for i := 0; i < minLength; i++ {
+				if sortedKeys[a].binary[i] == sortedKeys[b].binary[i] {
+					continue
+				}
+				return sortedKeys[a].binary[i] > sortedKeys[b].binary[i]
+			}
+
+			return true
+		})
+
+		for _, key := range sortedKeys {
+			// serialize key
+			if err := c.marshal(key.key, p, c.maxSliceLen); err != nil {
```

Comment:
We already have the serialized representation of the key -- we don't need to serialize it again

Reply:
You are absolutely right!